src/Contacts/src/CNLabeledValue.m

```diff
@@ -22,35 +22,69 @@
 @implementation CNLabeledValue
 
 - (instancetype)initWithLabel:(NSString *)label value:(id)value {
+    if (self = [super init]) {
+//        [self applyLabeledValueVariablesIdentifier: [[NSUUID alloc] init].UUIDString label:label value:value];
+        [self applyLabeledValueVariablesIdentifier: [[NSUUID alloc] init].UUIDString label:label value:value];
+    }
+    
+    return self;
+}
+
+- (instancetype)initWithIdentifier:(NSString *)identifier label:(NSString *)label value:(id)value {
+    if (self = [super init]) {
+        [self applyLabeledValueVariablesIdentifier:identifier label:label value:value];
+    }
+    
+    return self;
+}
+
+- (void)dealloc {
+    [_identifier release];
+    [_label release];
+    [_value release];
+    [super dealloc];
+}
+
+- (instancetype)applyLabeledValueVariablesIdentifier:(NSString *)identifier label:(NSString *)label value:(id)value {
+    if (value == nil) {
+        @throw [[NSException alloc] initWithName:@"CNPropertyInvalidTypeExpression" reason:@"Variable value cannot be null." userInfo:nil];
+    }
+    
+    _identifier = [identifier copy];
+    _label = [label copy];
+    _value = [value copy];
     return self;
 }
 
 + (instancetype)labeledValueWithLabel:(NSString *)label value:(id)value {
-    id new_class = [[[self class] alloc] init];
-    return new_class;
+    return [[[[self class] alloc] initWithLabel:label value:value] autorelease];
```

Comment:
There's no need to use `[self class]` here (and below), `self` is already the class (since this is a class method).

`[foo class]` is used for two reasons:
* To get a dynamic class of some _instace_, e.g. `[button class]` will likely evaluate to `NSButton`
* As a sort of "class literal": `[NSButton class]` will _also_ evaluate to `NSButton` (not to the metaclass!)

Why use `[NSButton class]` when you can say `NSButton` directly? That's an objc syntax quirk: you _cannot_ actually use `NSButton` (or other class names) directly in some contexts. You can _message_ it: `[NSButton foo]`, but you cannot e.g. store it in a variable or pass it as an argument:

```objc
Class buttonClass = NSButton;  // Nope, syntax error
[foo someMethodWithClass: NSButton];  // Nope

Class buttonClass = [NSButton class];  // Works
[foo someMethodWithClass: [NSButton class]];  // Works
```

This works because you are allowed to message the class, and `+ class` returns itself (as opposed to `- class`, the instance method, which returns the class of the instance).

Here, `self` is already a class and you can use it directly.

Reply:
So I changed `[[self class] alloc]` to `[self alloc]`. 

I don't get any warning when i do it to a class method (+), but if it were to do it to an instance method (-), I get a warning that `Instance method '-alloc' not found`.

Reply:
Of course; my point about `[self class]` not being needed only applies to class methods, because there `self` is the class itself; in an instance method you need to use either `[self class]` (or `object_getClass(self)`, or even just `isa` — but don't do that) or `CNLabeledValue` (or `[CNLabeledValue class]` if you need it for something else than sending it a message).

The difference between `[self class]` and `CNLabeledValue` in instance methods, of course, is that `[self class]` is the *dynamic* class of the object. If `self` is an instance of some subclass of `CNLabeledValue`, `[self class]` will be that subclass.

Reply:
Oh, I see you mean now.

So if the Apple documentation tells me that `labeledValueBySettingLabel:` "returns a labeled value object", should I be using `[CNLabeledValue alloc]` instead of `[[self class] alloc]` (for a instance method)?

Reply:
I think so, yes.

In this case it's not that important since no one will probably subclass `CNLabeledValue` anyway (KVOA will, if you know what that is, but it would hide that by making `[self class]` return the original `CNLabeledValue` as if nothing happened).